File: frontend/src/lib/api/officeStock.ts
import apiClient from './client';

export interface OfficeStockItem {
  id: number;
  empresa_id: number;
  office_id: number | null;
  sku_marketplace: string;
  sku_fornecedor: string | null;
  quantity: number;
  source_type: string;
  source_sales_order_id: number | null;
  source_sales_order_item_id: number | null;
  source_purchase_order_id: number | null;
  source_purchase_order_item_id: number | null;
  status: string;
  condition: string;
  received_at: string | null;
  rma_claim_id: number | null;
  created_at: string | null;
  consumed_by_sales_order_id: number | null;
  consumed_by_sales_order_item_id: number | null;
  consumed_at: string | null;
  office_nome?: string | null;
  empresa_nome?: string | null;
}

export const officeStockApi = {
  list: async (params: {
    empresa_id?: number;
    office_id?: number;
    sku?: string;
    status?: string;
    limit?: number;
    offset?: number;
  } = {}): Promise<OfficeStockItem[]> => {
    const search = new URLSearchParams();
    if (params.empresa_id != null) search.append('empresa_id', String(params.empresa_id));
    if (params.office_id != null) search.append('office_id', String(params.office_id));
    if (params.sku) search.append('sku', params.sku);
    if (params.status) search.append('status', params.status);
    search.append('limit', String(params.limit ?? 200));
    search.append('offset', String(params.offset ?? 0));
    const { data } = await apiClient.get<OfficeStockItem[]>(`/api/v1/office-stock?${search.toString()}`);
    return data;
  },

  getAvailableBySku: async (params: {
    empresa_id: number;
    sku_marketplace: string;
    office_id?: number;
  }): Promise<{ id: number; quantity: number; sku_fornecedor: string | null; source_sales_order_id: number | null; office_id: number | null; office_nome: string | null }[]> => {
    const search = new URLSearchParams();
    search.append('empresa_id', String(params.empresa_id));
    search.append('sku_marketplace', params.sku_marketplace);
    if (params.office_id != null) search.append('office_id', String(params.office_id));
    const { data } = await apiClient.get(`/api/v1/office-stock/available-by-sku?${search.toString()}`);
    return data;
  },

  consume: async (
    officeStockId: number,
    body: { quantity: number; sales_order_id: number; sales_order_item_id?: number }
  ): Promise<{ success: boolean; quantity_consumed: number; sales_order_id: number }> => {
    const { data } = await apiClient.post(`/api/v1/office-stock/${officeStockId}/consume`, body);
    return data;
  },
};

File: backend/app/api/v1/office_stock.py
"""
API: Stock em escritório (origem cancelamentos). Listar e consumir para reutilização.
"""
from typing import Any, Dict, List, Optional

from fastapi import APIRouter, Query, HTTPException
from pydantic import BaseModel

from app.services.office_stock_service import OfficeStockService

router = APIRouter(prefix="/office-stock", tags=["office-stock"])


class ConsumeStockBody(BaseModel):
    quantity: float
    sales_order_id: int
    sales_order_item_id: Optional[int] = None


@router.get("", response_model=List[Dict[str, Any]])
async def list_office_stock(
    empresa_id: Optional[int] = Query(None),
    office_id: Optional[int] = Query(None),
    sku: Optional[str] = Query(None),
    status: Optional[str] = Query(None, description="available | reserved | consumed"),
    limit: int = Query(200, ge=1, le=500),
    offset: int = Query(0, ge=0),
):
    """Lista stock em escritório (origem: cancelamentos). Filtros por empresa, escritório, SKU, status."""
    svc = OfficeStockService()
    try:
        return svc.list_stock(
            empresa_id=empresa_id,
            office_id=office_id,
            sku=sku,
            status=status,
            limit=limit,
            offset=offset,
        )
    finally:
        svc.close()


@router.get("/available-by-sku", response_model=List[Dict[str, Any]])
async def get_available_by_sku(
    empresa_id: int = Query(..., description="ID da empresa"),
    sku_marketplace: str = Query(..., description="SKU no marketplace"),
    office_id: Optional[int] = Query(None),
):
    """Devolve linhas de stock disponíveis para um SKU (para sugerir uso em novo pedido)."""
    svc = OfficeStockService()
    try:
        return svc.get_available_by_sku(
            empresa_id=empresa_id,
            sku_marketplace=sku_marketplace,
            office_id=office_id,
        )
    finally:
        svc.close()


@router.post("/{office_stock_id}/consume")
async def consume_stock(office_stock_id: int, body: ConsumeStockBody):
    """Consome quantidade de uma linha de office_stock e associa ao pedido (reutilização)."""
    svc = OfficeStockService()
    try:
        result = svc.consume_stock(
            office_stock_id=office_stock_id,
            quantity=body.quantity,
            sales_order_id=body.sales_order_id,
            sales_order_item_id=body.sales_order_item_id,
        )
        if not result.get("success"):
            raise HTTPException(status_code=400, detail=result.get("error", "Erro ao consumir stock"))
        return result
    finally:
        svc.close()
